pkg/modifycontroller/modify_volume.go

```diff
@@ -0,0 +1,178 @@
+		return ctrl.controllerModifyVolumeWithTarget(pvc, pv, vacObj.(*storagev1alpha1.VolumeAttributesClass), pvcSpecVacName)
+	} else {
+		// Mark pvc.Status.ModifyVolumeStatus as pending
+		pvc, err = ctrl.markControllerModifyVolumeStatus(pvc, v1.PersistentVolumeClaimModifyVolumePending, nil)
+		return pvc, pv, err, false
+	}
+}
+
+// func controllerModifyVolumeWithTarget trigger the CSI ControllerModifyVolume API call
+// and handle both success and error scenarios
+func (ctrl *modifyController) controllerModifyVolumeWithTarget(
+	pvc *v1.PersistentVolumeClaim,
+	pv *v1.PersistentVolume,
+	vacObj *storagev1alpha1.VolumeAttributesClass,
+	pvcSpecVacName *string) (*v1.PersistentVolumeClaim, *v1.PersistentVolume, error, bool) {
+	var err error
+	pvc, pv, err = ctrl.callModifyVolumeOnPlugin(pvc, pv, vacObj)
+	if err == nil {
+		klog.V(4).Infof("Update volumeAttributesClass of PV %q to %s succeeded", pv.Name, *pvcSpecVacName)
+		// Record an event to indicate that modify operation is successful.
+		ctrl.eventRecorder.Eventf(pvc, v1.EventTypeNormal, util.VolumeModifySuccess, fmt.Sprintf("external resizer modified volume %s with vac %s successfully ", pvc.Name, vacObj.Name))
+		return pvc, pv, nil, true
+	} else {
+		if status, ok := status.FromError(err); ok && status.Code() == codes.InvalidArgument {
+			// Mark pvc.Status.ModifyVolumeStatus as infeasible
+			pvc, err = ctrl.markControllerModifyVolumeStatus(pvc, v1.PersistentVolumeClaimModifyVolumeInfeasible, err)
+		} else {
+			ctrl.updateConditionBasedOnError(pvc, err)
+			if !isFinalError(err) {
+				// update conditions and cache pvc as uncertain
+				pvcKey, err := cache.MetaNamespaceKeyFunc(pvc)
+				if err != nil {
+					return pvc, pv, err, false
+				}
+				ctrl.uncertainPVCs[pvcKey] = *pvc
+			} else {
+				ctrl.removePVCFromModifyVolumeUncertainCache(pvc)
+			}
+		}
+		// Record an event to indicate that modify operation is failed.
+		ctrl.eventRecorder.Eventf(pvc, v1.EventTypeWarning, util.VolumeModifyFailed, err.Error())
+		return pvc, pv, err, false
+	}
+}
+
+func (ctrl *modifyController) callModifyVolumeOnPlugin(
+	pvc *v1.PersistentVolumeClaim,
+	pv *v1.PersistentVolume,
+	vac *storagev1alpha1.VolumeAttributesClass) (*v1.PersistentVolumeClaim, *v1.PersistentVolume, error) {
+	err := ctrl.modifier.Modify(pv, vac.Parameters)
+
+	if err != nil && isFinalError(err) {
+		// Mark pvc.Status.ModifyVolumeStatus as infeasible
+		pvc, markModifyVolumeInfeasibleError := ctrl.markControllerModifyVolumeStatus(pvc, v1.PersistentVolumeClaimModifyVolumeInfeasible, err)
+		if markModifyVolumeInfeasibleError != nil {
+			return pvc, pv, fmt.Errorf("modify volume failed in controller with %v but failed to update PVC %s with: %v", err, pvc.Name, markModifyVolumeInfeasibleError)
+		}
+	}
+
+	if err != nil {
+		return pvc, pv, err
+	}
+
+	pvc, pv, err = ctrl.markControllerModifyVolumeCompleted(pvc, pv)
+	return pvc, pv, nil
+}
+
+func isFinalError(err error) bool {
```

Comment:
Can we reuse this function between both resize and modify controllers?

Reply:
Moved to util package.